src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1081,6 +1082,74 @@ private void ApplyTextDocumentChange(DocumentId documentId, SourceText newText)
             }
         }
 
+        protected override void ApplyDocumentInfoChanged(DocumentId id, DocumentInfo updatedInfo)
+        {
+            var document = CurrentSolution.GetDocument(id);
+
+            FailIfDocumentInfoChangesNotSupported(document, updatedInfo);
+
+            if (document.Name != updatedInfo.Name)
+            {
+                GetProjectData(updatedInfo.Id.ProjectId, out var _, out var project);
+
+                var projectItemForDocument = project.FindItemByPath(document.FilePath, StringComparer.OrdinalIgnoreCase);
+
+                if (projectItemForDocument == null)
+                {
+                    // TODO: The document might actually belong to a Shared Project.
+                    return;
+                }
+
+                // Must save the document first for things like Breakpoints to be preserved.
+                projectItemForDocument.Save();
+
+                var uniqueName = projectItemForDocument.Collection.GetUniqueName(
+                    Path.GetFileNameWithoutExtension(updatedInfo.Name),
+                    Path.GetExtension(updatedInfo.Name));
+
+                // By setting this property, Visual Studio will perform the file rename, which 
+                // will cause the workspace's current solution to update and will fire the 
+                // necessary workspace changed events.
+                projectItemForDocument.Name = uniqueName;
+            }
+        }
+
+        /// <summary>
+        /// The VisualStudioWorkspace currently supports only a subset of DocumentInfo changes.
+        /// </summary>
+        private void FailIfDocumentInfoChangesNotSupported(CodeAnalysis.Document document, DocumentInfo updatedInfo)
+        {
+            if (document.SourceCodeKind != updatedInfo.SourceCodeKind)
+            {
+                throw new InvalidOperationException(
+                    $"This Workspace does not support changing a document's {nameof(document.SourceCodeKind)}.");
+            }
+
+            if (document.FilePath != updatedInfo.FilePath)
```

Comment:
Is extracting out the "if two values are different, throw exception" to a local function make this a bit easier to read? It'd then compress all the actual checks making it easier to see if you actualy remembered everything.

Reply:
(the code is fine, it's just scrolling on a laptop screen which makes reviewing harder....)

Reply:
Such a local function would have to take objects, right? I think you should just get a bigger laptop. 😆

Reply:
OK, tagging @jinujoseph to get me a bigger laptop. :smile: